BZ1973326: Update firstboot multipath doc (#34462)
openshift/openshift-docs: modules/rhcos-enabling-multipath.adoc

[id="rhcos-enabling-multipath_{context}"]
= Enabling multipathing with kernel arguments on {op-system}

{op-system} supports multipathing on the primary disk, allowing stronger resilience to hardware failure to achieve higher host availability.

You can enable multipathing at installation time for nodes that were provisioned in {product-title} 4.8 or later. While post-installation support is available by activating multipathing via the machine config, enabling multipathing during installation is recommended.

In setups where any I/O to non-optimized paths results in I/O system errors, you must enable multipathing at installation time.

[IMPORTANT]
====
On IBM Z and LinuxONE, you can enable multipathing only if you configured your cluster for it during installation. For more information, see "Installing {op-system} and starting the {product-title} bootstrap process" in _Installing a cluster with z/VM on IBM Z and LinuxONE_.
====
// Add xref once it's allowed.

.Prerequisites
* You have a running {product-title} cluster that uses version 4.8 or later.
* You are logged in to the cluster as a user with administrative privileges.

.Procedure

> . To enable multipathing at installation time, append the kernel arguments by entering the following command:

Reviewer: There's two facets actually for a fully multipathed installation:
1. We want `coreos-installer install` to write to the multipathed device. This means that users should use e.g. `coreos-installer install /dev/mapper/mpatha` or using a WWN-named link under `/dev/disk/by-id`. That requires turning on multipath in the installation environment, which could be done using `rd.multipath=default` in the PXE/ISO boot, or just manually calling e.g. `mpathconf --enable` and `systemctl start multipathd.service` before calling `coreos-installer install`.
2. Appending kargs to the `coreos-installer install` command so that the installed system itself will use multipath right from the first boot. This part you already have.

Author: OK, thanks for the details. I've added an initial step based on whether you are doing an ISO/PXE boot or not. But I'm not sure if I should incorporate this part of your comment: 
>  This means that users should use e.g. `coreos-installer install /dev/mapper/mpatha` or using a WWN-named link under `/dev/disk/by-id`.

If so, how would you recommend doing so? As a prerequisite like the following perhaps?

```
.Prerequisites
* You are using either `coreos-installer install /dev/mapper/mpatha` or a World Wide Name (WWN) link under `/dev/disk/by-id`.
```

+
[source,terminal]
----
$ coreos-installer install --append-karg rd.multipath=default --append-karg root=/dev/disk/by-label/dm-mpath-root
----

. Check that the kernel argument worked by going to one of the worker nodes and listing

> the kernel command line arguments (in `/proc/cmdline` on the host):

Author: This is taken from the post-install multipath procedure. Is it relevant here or is there another way that the user can verify that the command worked?

Reviewer: Yes, it's still relevant.

====
// Add xref once it's allowed.

The following procedure enables multipath at installation time and appends kernel arguments to the `coreos-installer install` command so that the installed system itself will use multipath beginning from the first boot.

.Prerequisites
* You have a running {product-title} cluster that uses version 4.8 or later.
* You are logged in to the cluster as a user with administrative privileges.

.Procedure

. To enable multipath and start the `multipathd` daemon, run the following command:
+
[source,terminal]
----
$ mpathconf --enable && systemctl start multipathd.service
----
** Optional: If booting the PXE or ISO, you can instead enable multipath by adding `rd.multipath=default` from the kernel command line.

. Edit the `/etc/multipath.conf` file to target the multipath device directly for installation, either on the kernel command line by using `coreos.inst.install_dev` or interactively by invoking the `coreos-installer` program.
+

> You can configure multipath so that the names of the multipath devices are of the form `/dev/mapper/mpath` `_n_`. Without this setting, the names of the multipath devices are aliased to a World Wide Name (WWN) symlink, such as `/dev/disk/by-label`.

Reviewer: Unfortunately, we have no easy way right now for customers to provide custom multipath configs to the installed node on first boot. So they have to stay with the default configs emitted by `mpathconf --enable`.

If there's a single multipath device, then it *should* reliably be `/dev/mapper/mpatha`. If there are multiple multipath devices, or if they want to be more explicit, then using the `/dev/disk/by-id/wwn-<WWN>` symlinks would be best.

Author: Forgive my ignorance, please, as I try to understand the steps that the user would take here.

Can you provide an example output that the user would see after running step 1 (`$ mpathconf --enable && systemctl start multipathd.service`)? I'm thinking that could help with step two, where I've changed it to reflect your wording by saying, "Use the default configs emitted by `mpathconf --enable`".

Author: Thanks, @jlebon, and pardon my ignorance in this process, please. 😁 

I've made another round of updates to hopefully give clearer directions on what users need to do to append kargs using `coreos-installer install`. So now in Step 2, I've broken it out for instructions to do either: A) A single mpath device, or B) Multiple mpath devices. I also left a note in there that they can use `coreos.inst.install_dev`, though I'm not certain how. If you have a suggestion for what step(s) the user should take for this, let me know and I'll include it. Thanks!

. Append the kernel arguments by entering the following command:
+
[source,terminal]
----

> $ coreos-installer install --append-karg rd.multipath=default --append-karg root=/dev/disk/by-label/dm-mpath-root

Author: Is `by-label` sub dir okay here to stay in line with other examples in this section? Or should it really be `by-id`?

Reviewer: This path is good. Once RHCOS is installed, the root partition and label exist so it's valid. Before installing RHCOS, it's just a random disk, so they have to use e.g. `/dev/mapper/` or `/dev/disk/by-id/wwn-...` to refer to the multipathed disk when installing. Which, speaking of, should this example include the target device to be clearer? E.g.

```
coreos-installer install /dev/disk/by-id/wwn-0xx194e957fcedb4841 --append-karg rd.multipath=default --append-karg root=/dev/disk/by-label/dm-mpath-root
```

Author: Sure, if that makes sense to you it's good with me. 😁  I changed the lead-in wording and the code block slightly to indicate that this is an example with a replaceable value. Do my edits make sense?

In setups where any I/O to non-optimized paths results in I/O system errors, you must enable multipathing at installation time.

[IMPORTANT]
====
On IBM Z and LinuxONE, you can enable multipathing only if you configured your cluster for it during installation. For more information, see "Installing {op-system} and starting the {product-title} bootstrap process" in _Installing a cluster with z/VM on IBM Z and LinuxONE_.
====
// Add xref once it's allowed.

.Prerequisites
* You have a running {product-title} cluster that uses version 4.8 or later.
* You are logged in to the cluster as a user with administrative privileges.

.Procedure

* To enable the multipath service at installation time:
+
. Turn on multipath in the installation environment.
+
** For a PXE or ISO boot, add the `rd.multipath=default` kernel argument to your `MachineConfig` object file. For information about setting kernel arguments, see "Adding kernel arguments to nodes".
+

> ** For a non-PXE/ISO boot, run the following commands:

Reviewer: How about something like:

- Turn on multipath using `mpathconf --enable && systemctl start multipathd.service` (for more information, see the [RHEL 8 multipath documentation](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/configuring_device_mapper_multipath/assembly_setting-up-dm-multipath-configuring-device-mapper-multipath#proc_basic-multipath-setup-setting-up-dm-multipath)). If booting the PXE or ISO, you may also instead add `rd.multipath=default` to turn on multipath from the kernel command-line.
- This will allow you to target the multipath device directly for installation using e.g. `/dev/mapper/mpatha` or a WWN-named symlink in `/dev/disk/by-id`. As usual, the device path can be specified on the kernel command-line (i.e. via `coreos.inst.install_dev`) or interactively (i.e. when invoking `coreos-installer`).

? Not sure if it's too wordy.

Author: @jlebon Made some more tweaks by incorporating the RHEL 8 docs. Thanks for the link. I need to make it task-based, so that's why it's broken into what seems like so many steps. Hoping I got the ordering and context right but please let me know. 🙇

====
// Add xref once it's allowed.

The following procedure enables multipath at installation time and appends kernel arguments to the `coreos-installer install` command so that the installed system itself will use multipath beginning from the first boot.

.Prerequisites
* You have a running {product-title} cluster that uses version 4.8 or later.
* You are logged in to the cluster as a user with administrative privileges.

.Procedure

. To enable multipath and start the `multipathd` daemon, run the following command:
+
[source,terminal]
----
$ mpathconf --enable && systemctl start multipathd.service
----
** Optional: If booting the PXE or ISO, you can instead enable multipath by adding `rd.multipath=default` from the kernel command line.

. Use the default configs emitted by `mpathconf --enable` from the previous step to specify the multipath device directly for installation, either on the kernel command line by using `coreos.inst.install_dev` or interactively by invoking the `coreos-installer` program.
+

> For a single multipath device, use the path `/dev/mapper/mpatha`.

Author: Is there a `coreos-installer` command example we could provide here?

// Module included in the following assemblies:
//
// * installing/installing_bare_metal/installing-bare-metal.adoc
// * installing/installing_bare_metal/installing-bare-metal-network-customizations.adoc
// * installing/installing_bare_metal/installing-restricted-networks-bare-metal.adoc

[id="rhcos-enabling-multipath_{context}"]
= Enabling multipathing with kernel arguments on {op-system}

{op-system} supports multipathing on the primary disk, allowing stronger resilience to hardware failure to achieve higher host availability.

> You can enable multipathing at installation time for nodes that were provisioned in {product-title} 4.8 or later. While post-installation support is available by activating multipathing via the machine config, enabling multipathing during installation is recommended.

Reviewer 2: {product-version} ?

Author: Same comment as above, leaving 4.8 hardcoded.

[IMPORTANT]
====
On IBM Z and LinuxONE, you can enable multipathing only if you configured your cluster for it during installation. For more information, see "Installing {op-system} and starting the {product-title} bootstrap process" in _Installing a cluster with z/VM on IBM Z and LinuxONE_.
====
// Add xref once it's allowed.

The following procedure enables multipath at installation time and appends kernel arguments to the `coreos-installer install` command so that the installed system itself will use multipath beginning from the first boot.

.Prerequisites
* You have a running {product-title} cluster that uses version 4.8 or later.
* You are logged in to the cluster as a user with administrative privileges.

.Procedure

. To enable multipath and start the `multipathd` daemon, run the following command:
+
[source,terminal]
----
$ mpathconf --enable && systemctl start multipathd.service
----
** Optional: If booting the PXE or ISO, you can instead enable multipath by adding `rd.multipath=default` from the kernel command line.

. Append the kernel arguments by invoking the `coreos-installer` program:
+

> * For a single multipath device, use the path `/dev/mapper/mpatha`, for example:

Reviewer: Maybe instead, something like:

```suggestion
* If there is only one multipath device connected to the machine, it should be available at path `/dev/mapper/mpatha`. For example:
```

Author: Resolved

+
[source,terminal]
----
$ coreos-installer install /dev/mapper/mpatha \ <1>
--append-karg rd.multipath=default \

> --append-karg root=/dev/mapper/dm-mpath-root

Author: Please double-check if these paths are correct for this example.

Reviewer: ```suggestion
$ coreos-installer install /dev/mapper/mpatha \ <1>
--append-karg rd.multipath=default \
--append-karg root=/dev/disk/by-label/dm-mpath-root
```

The `root=` karg should always be `/dev/disk/by-label/dm-mpath-root`. Here we're talking about the path to the overall block device itself onto which all of RHCOS is installed (e.g. on top of which the boot partition, EFI and rootfs partitions are created).

Author: Thanks for the explanation, resolved.

----
<1> Indicates the path of the single multipathed device.
+

> * For multiple multipath devices or to be more explicit, instead of using a path such as `/dev/mapper/mpatha`, it is recommended to use the World Wide Name (WWN) symlink available in `/dev/disk/by-id`. For example:

Reviewer: And then here

```suggestion
* If there are multiple multipath devices or to be more explicit, instead of using `/dev/mapper/mpatha`, it is recommended to use the World Wide Name (WWN) symlink available in `/dev/disk/by-id`. For example:
```

Author: Resolved

> * For multiple multipath devices, or to be more explicit, it is recommended that the names of the multipath devices are aliased to a World Wide Name (WWN) symlink when you append the kernel arguments , such as `/dev/disk/by-label`. For example:

Reviewer: ```suggestion
* For multiple multipath devices, or to be more explicit, instead of using e.g. `/dev/mapper/mpatha`, it is recommended to use the World Wide Name (WWN) symlink available in `/dev/disk/by-id`. For example:
```

Author: Resolved

+
[source,terminal]
----

> $ coreos-installer install /dev/disk/by-id/<wwn_ID> \ <1>

Reviewer: Should this be

```suggestion
$ coreos-installer install /dev/disk/by-id/wwn-<wwn_ID> \ <1>
```

instead, and then the `<1>` below is just `0xx194e957fcedb4841` for the WWN ID?

Author: Resolved

--append-karg rd.multipath=default \
--append-karg root=/dev/disk/by-label/dm-mpath-root
----
<1> Indicates the WWN ID of the target multipathed device. For example, `wwn-0xx194e957fcedb4841`.
+

> Alternatively, you can append the kernel arguments on the kernel command line by using `coreos.inst.install_dev`.

Author: Should we describe how this is done as an optional step?

Reviewer: Yeah, not sure where it belongs. Maybe better to just link to the docs where we detail the "install via coreos.inst kargs" path? E.g. something like:

> This symlink can also be used as the `coreos.inst.install_dev` kernel argument when using the [automated installation procedure](link to relevant page).

?

Author: Sure, that sounds good. We currently have technical limitations to our docs that don't allow xrefs from a module such as this, but I have referenced the title of the relevant doc section.

> This symlink can also be used as the `coreos.inst.install_dev` kernel argument when using special `coreos.inst.*` arguments to direct the live installer. For more information, see "Creating {op-system-first} machines".

Author: Definitely! It also made me realize that we changed the title of the xref in 4.8, so updated that here as well. Thanks!
